server/src/main/java/io/pravega/schemaregistry/service/SchemaRegistryService.java

```diff
@@ -963,16 +973,51 @@ private Boolean canReadChecker(SchemaInfo schema, GroupProperties prop, List<Sch
      * @return Map of group id to version that identifies the schema in the group.
      */
     public CompletableFuture<Map<String, VersionInfo>> getSchemaReferences(String namespace, SchemaInfo schemaInfo) {
-        SchemaInfo schema = normalizeSchemaBinary(schemaInfo);
+        NormalizedForm normalizedForm = normalizeSchemaBinary(schemaInfo);
+        BigInteger fingerprint = normalizedForm.getFingerprint();
 
-        return store.getGroupsUsing(namespace, schema)
+        return store.getGroupsUsing(namespace, fingerprint, normalizedForm::compareNormalized)
                     .thenCompose(groups -> Futures.allOfWithResults(
-                            groups.stream().collect(Collectors.toMap(x -> x, x ->
-                                    Futures.exceptionallyExpecting(store.getSchemaVersion(namespace, x, schema),
-                                            e -> Exceptions.unwrap(e) instanceof StoreExceptions.DataNotFoundException, EMPTY_VERSION))))
+                            groups.stream().collect(Collectors.toMap(x -> x, x -> {
+                                return Futures.exceptionallyExpecting(store.getSchemaVersion(namespace, x, fingerprint, 
+                                        normalizedForm::compareNormalized),
+                                        e -> Exceptions.unwrap(e) instanceof StoreExceptions.DataNotFoundException, EMPTY_VERSION);
+                            })))
                                                   .thenApply(result -> {
                                                       return result.entrySet().stream().filter(x -> !x.getValue().equals(EMPTY_VERSION))
                                                                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
                                                   }));
     }
+    
+    @Data
+    static class NormalizedForm {
+        private final SchemaInfo normalized;
+        private final Object parsedSchema;
+        private final BigInteger fingerprint;
+        private NormalizedForm(SchemaInfo schemaInfo, Object parsedSchema) {
+            this.normalized = schemaInfo;
+            this.parsedSchema = parsedSchema;
+            this.fingerprint = HashUtil.getFingerprint(schemaInfo.getSchemaData().array());
+        }
+
+        boolean compareNormalized(SchemaInfo schemaInfo) {
```

Comment:
One potential problem here is that the `SchemaInfo` is not normalized, so it is error prone. Should we have a derived type to indicate that the object is actually normalized? Maybe replace this class with `NormalizedSchemaInfo` and implement a comparison method (`compareTo`) for this instance, would it work?

Reply:
If you look at the method's implementation, it normalizes schema info and then compares it with the normalized form. 

`NormalizedForm` is the class which parses and holds the schema in its normalized form. we could do it in two steps as you suggest too.